### DISCUSSION TITLE
W5500 MAC Adress handling

### DISCUSSION THREAD

**Discussion by Ing-Dom** [earlephilhower/arduino-pico#1720] [Q&A] [answered]
first of all, thank you for this new LAN support :)

While it puzzled me where the MAC-Adress comes from in the W5500 example i solved that with digging through the ocde.

https://github.com/earlephilhower/arduino-pico/blob/master/libraries/lwIP_Ethernet/src/LwipIntfDev.h#L278C1-L282C6

I cannot find details to
cyw43_hal_generate_laa_mac
anywhere - is the output reproducable? On what "source" is the mac generated?
Will the same device always get the same MAC?
Will identical deviecs get the same MAC ?
(I used a part of the W25Qs id until now..)

Second - there seems to be no possibility to get the mac over the Wiznet5500lwIP object... shouldn't be there the possibility?

**Comment by earlephilhower** [comment, marked as answer]
See https://github.com/raspberrypi/pico-sdk/blob/6a7db34ff63345a7badec79ebea3aaef1712f374/src/rp2_common/pico_cyw43_driver/cyw43_driver.c#L132 for the MAC generation.  It uses the flash ID and should be unique. It's the same code used on the WiFi driver.  Each interface also gets adjusted so if you have 2 network drivers on the same device they won't have identical MACs 

As for getting the MAC, that seems reasonable.  Maybe you'd like to do a PR?  I think the LeipIntfDev would be the spot to add the accessor/setter.

**Reply by Ing-Dom** [reply]
Thanks. I always forget that the PICO-SDK is distributed in binary form with your core. Thats why I didn't find it..
The solution looks good.

I'll do a PR  for the MAC getter/setter.

**Comment by Ing-Dom** [comment]
I found out you can access the mac through the netif.

eth.getNetIf()->hwaddr;

Do you think there is still a need on the LeipIntfDev level?

**Reply by earlephilhower** [reply]
Nope, I think we're okay without it.  Thanks for checking!
